### PR TITLE
Test/fix scoping rules for pattern locals

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@gafter, @jaredpar, @dotnet/roslyn-compiler Please review.
CC @MattGertz for ask mode approval.

**Comment by MattGertz** [comment]
@AlekseyTs Which issue does this fix?

**Comment by AlekseyTs** [comment]
@MattGertz Fixes #8816.

File: src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs [hunk 1]
```diff
@@ -94,7 +94,7 @@ internal static void BindAttributeTypes(ImmutableArray<Binder> binders, Immutabl
 
         internal CSharpAttributeData GetAttribute(AttributeSyntax node, NamedTypeSymbol boundAttributeType, DiagnosticBag diagnostics)
         {
-            var boundAttribute = BindAttribute(node, boundAttributeType, diagnostics);
+            var boundAttribute = new PatternVariableBinder(node, this).BindAttribute(node, boundAttributeType, diagnostics);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs]: Does this have the desired impact on the semantic model? Or do we even care about that?

**Comment by MattGertz** [comment]
Thanks.  Approved for 2.0 pending code reviews.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 2]
```diff
@@ -339,7 +339,8 @@ internal BoundExpression BindValueAllowArgList(ExpressionSyntax node, Diagnostic
             ParameterSymbol parameter,
             EqualsValueClauseSyntax defaultValueSyntax)
         {
-            return new LocalScopeBinder(this.WithContainingMemberOrLambda(parameter.ContainingSymbol).WithAdditionalFlags(BinderFlags.ParameterDefaultValue));
+            return new LocalScopeBinder(this.WithContainingMemberOrLambda(parameter.ContainingSymbol).WithAdditionalFlags(BinderFlags.ParameterDefaultValue)).
+                   WithPatternVariablesIfAny(defaultValueSyntax.Value);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Does this have the desired impact on the semantic model? Or do we even care about that?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 3]
```diff
@@ -2644,12 +2645,8 @@ private static bool IsNegativeConstantForArraySize(BoundExpression expression)
...
             DiagnosticBag diagnostics)
         {
-            // Handle scoping for possible pattern variables declared in the initializer
-            PatternVariableBinder patBinder = (initializerArgumentListOpt != null)
-                ? new PatternVariableBinder(initializerArgumentListOpt, initializerArgumentListOpt.Arguments, this)
-                : null;
-            var result = (patBinder ?? this).BindConstructorInitializerCore(initializerArgumentListOpt, constructor, diagnostics);
-            return patBinder?.WrapWithPatternVariables(result) ?? result;
+            var result = BindConstructorInitializerCore(initializerArgumentListOpt, constructor, diagnostics);
+            return WrapWithVariablesIfAny(result);
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Does this have the desired impact on the semantic model? Or do we even care about that?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 2 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Oh, I see where that is done later.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 3 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Oh, I see where that is done later.

File: src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs]: Oh, I see where that is done later.

**Comment by gafter** [comment]
LGTM

**Comment by jaredpar** [comment]
:+1:

**Comment by VSadov** [comment]
LGTM

**Pull request merged**